File: SchoolWebApp-API/Project.Core/Entities/Finance/BudgetMaster.cs
using SchoolWebApp.Core.Entities.Academics;
using SchoolWebApp.Core.Entities.Shared;
using System.ComponentModel.DataAnnotations;

namespace SchoolWebApp.Core.Entities.Finance
{
    public enum BudgetMasterStatus
    {
        Draft = 0,
        Open = 1,
        Closed = 2
    }

    public class BudgetMaster : Base
    {
        [Required]
        [StringLength(255)]
        public required string Name { get; set; }

        [StringLength(50)]
        public string? Code { get; set; }

        [StringLength(500)]
        public string? Description { get; set; }

        public int AcademicYearId { get; set; }
        public AcademicYear? AcademicYear { get; set; }

        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }

        public BudgetMasterStatus Status { get; set; } = BudgetMasterStatus.Draft;

        public List<Budget> Budgets { get; set; } = new();
    }
}

File: SchoolWebApp-Client/src/app/finance/models/budget-master.ts
import {ResourceModel} from '@/core/models/ResourceModel';

export enum BudgetMasterStatus {
    Draft = 0,
    Open = 1,
    Closed = 2
}

export class BudgetMaster extends ResourceModel<BudgetMaster> {
    public name?: string;
    public code?: string;
    public description?: string;
    public academicYearId?: number;
    public academicYearName?: string;
    public startDate?: string;
    public endDate?: string;
    public status?: BudgetMasterStatus;
    public budgetCount?: number;

    constructor(model?: Partial<BudgetMaster>) {
        super(model);
    }
}

export enum BudgetAmendmentStatus {
    Pending = 0,
    Approved = 1,
    Rejected = 2
}

export class BudgetAmendmentLine {
    public id?: number;
    public budgetAmendmentId?: number;
    public accountId?: number;
    public accountCode?: string;
    public accountName?: string;
    public previousAmount?: number;
    public newAmount?: number;
    public delta?: number;
    public notes?: string;
}

export class BudgetAmendment extends ResourceModel<BudgetAmendment> {
    public budgetId?: number;
    public budgetName?: string;
    public referenceNumber?: string;
    public amendmentDate?: string;
    public reason?: string;
    public status?: BudgetAmendmentStatus;
    public approvedById?: number;
    public approvedByName?: string;
    public approvedDate?: string;
    public lines!: BudgetAmendmentLine[];

    constructor(model?: Partial<BudgetAmendment>) {
        super(model);
        if (!this.lines) this.lines = [];
    }
}
